CIP-0037/CIP-0037.md

```diff
@@ -0,0 +1,338 @@
+interpreted. These bits are used to unambiguously represent the underlying element:
+* The most significant bit, when set, indicates that the point is in compressed form.
+* The second-most significant bit indicates that the point is at infinity. If this bit is set, the remaining bits of
+  the group element's encoding should be set to zero.
+* The third-most significant bit is set if (and only if) this point is in compressed form and it is not the point at
+  infinity and its y-coordinate is the lexicographically largest of the two associated with the encoded x-coordinate.
+
+We include the serialisation of the generator of G1 and the generator of G2:
+* generator G1:
+```
+[151, 241, 211, 167, 49, 151, 215, 148, 38, 149, 99, 140, 79, 169, 172, 15, 195, 104, 140, 79, 151, 116, 185, 
+5, 161,78, 58, 63, 23, 27, 172, 88, 108, 85, 232, 63, 249, 122, 26, 239, 251, 58, 240, 10, 219, 34, 198, 187]
+```
+* generator G2:
+```
+[147, 224, 43, 96, 82, 113, 159, 96, 125, 172, 211, 160, 136, 39, 79, 101, 89, 107, 208, 208, 153, 32, 182, 
+26, 181, 218, 97, 187, 220, 127, 80, 73, 51, 76, 241, 18, 19, 148, 93, 87, 229, 172, 125, 5, 93, 4, 43, 126, 
+2, 74, 162, 178, 240, 143, 10, 145, 38, 8, 5, 39, 45, 197, 16, 81, 198, 228, 122, 212, 250, 64, 59, 2, 180, 
+81, 11, 100, 122, 227, 209, 119, 11, 172, 3, 38, 168, 5, 187, 239, 212, 128, 86, 200, 193, 33, 189, 184]
+```
+
+#### An important note on GT elements
+We intentionally limit what can be done with the GT element. In fact, the only way that one can generate a 
+GT element is using the pairing function. Then these elements can only be used for operations among 
+them (mult) and a final equality check. In other words, GT elements are only there to eventually
+perform some equality checks. We thought of including the inverse
+function to allow division, but given that we simply allow for equality checks, if one needs to divide by `A`, 
+then `A` can simply move to the other side of the equality sign. These limitations allow us for a performance 
+trick, which is also used for the verification of BLS signatures. In a nutshell, a pairing is divided into two 
+operations: (i) Miller loop, and (ii) final exponentiation. Both are expensive, but what we can do is first 
+compute the miller loop, which already maps two points from G_1 and G_2 to G_T. Then we can use this result 
+to perform the operations over these points (mults). Finally, when we want to check for equality, we invert 
+one of the two points, and multiply the result by the other point. Only now we compute the final exponentiation,
+and we verify that the result is the identity element. In other words, the equality check function, `eq`, first
+computes an inverse, then a multiplication, and finally a final exponentiation.
+                                         
+For this reason it is important to limit what can be done with GTElements, as the pairing really is not the full 
```

Comment:
So essentially, you'd add `BLS12_381_full_pairing`? Shouldn't `BLS12_381_pairing` just have a different name then?

Reply:
Yes, right. If there is the possibility that in the future we would add the full pairing, we should name it differently here. Maybe something like `BLS12_381_ppairing_ml` (standing for partial pairing miller loop), and then have clear documentation of the function.

Reply:
@michaelpj Thoughts on this? Should we try to make a complete API or should we add it in another Plutus language version in the future if someone needs it?

Reply:
bump

Reply:
It's hard to say in advance. I think the efficiency arguments are compelling, but it would indeed be good to identify clearly what the functions that we are adding do, e.g. naming and documenting them appropriately. That will also reduce confusion if we do add more functions later.